### PR TITLE
Created Booking Model #2

### LINKED ISSUE #2: Create Booking Model (M)
Draft version:
Booking
- id
-serialNumber (scooter)
- accountId
- startDate
- endDate
- payment (SUCCESS/FAILURE)
- cost?

### PR DESCRIPTION
**Pull request by birleadrian**
I've created the Booking Model #2 using Lombok to generate Getters , Setters and Constructors.

### CODE REVIEW

File: src/main/java/com/bootcamp/demo/models/Booking.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import java.util.Date;
+
+
+@Getter
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
```
Comment (dbernatt) [inline src/main/java/com/bootcamp/demo/models/Booking.java]: You can use `@Data` instead of `@Getter`,` @Setter`. 
You can read about `@Data` annotation here -> [lombok @data](https://projectlombok.org/features/Data)

File: src/main/java/com/bootcamp/demo/models/Booking.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+public class Booking {
+
+    private long id;
+    private long serialNumber;
+    private long accountId;
```
Comment (dbernatt) [inline src/main/java/com/bootcamp/demo/models/Booking.java]: I think account id should be an UUID type.

File: src/main/java/com/bootcamp/demo/models/PaymentStatus.java [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package com.bootcamp.demo.models;
+
+public enum PaymentStatus {
+    SUCCESS,
+    FAILURE
```
Comment (dbernatt) [inline src/main/java/com/bootcamp/demo/models/PaymentStatus.java]: [java enum serialization and deserialization](http://www.vineetmanohar.com/2010/01/3-ways-to-serialize-java-enums/)
You need to add a constructor where you specify the value of each enum category. You can read about this on the link above :).

> What does serialization and deserialization of enum mean? Serialization involves converting the enum to a value (usually primitive or a String) which can be easily stored on the disk or a database. Deserialization involves reading a stored value and converting it back to an enum.

File: src/main/java/com/bootcamp/demo/security/SecurityConfig.java [hunk 4]
```diff
@@ -29,7 +29,7 @@ public void configure(final HttpSecurity http) throws Exception {
     @Override
     protected void configure(final AuthenticationManagerBuilder auth) throws Exception {
         final var inMemoryAuth = auth.inMemoryAuthentication();
-        inMemoryAuth.withUser("test.team@gmail.com").password("{noop}xxx").roles("USER");
+        inMemoryAuth.withUser("").password("{noop}xxx").roles("USER");
```
Comment (dbernatt) [inline src/main/java/com/bootcamp/demo/security/SecurityConfig.java]: The username and password needs to be changed?

**Review by nymphadora-tonks-team — COMMENTED** [review]
We would like not to use lombok. You can find some details regarding the reasons for this in https://medium.com/@vgonzalo/dont-use-lombok-672418daa819

**Comment by birleadrian** [comment]
I think for ID we can use an UUID cause it's immutable.

**Review by dbernatt — CHANGES_REQUESTED** [review]
I reviewed your code. Please check out my comments.

File: src/main/java/com/bootcamp/demo/models/Booking.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
+package com.bootcamp.demo.models;
+import java.util.Date;
+import java.util.Objects;
+
+
+public class Booking {
```
Comment (dbernatt) [inline src/main/java/com/bootcamp/demo/models/Booking.java]: Class should be final

**Review by dbernatt — CHANGES_REQUESTED** [review]

File: src/main/java/com/bootcamp/demo/models/Booking.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    private long id;
+    private long serialNumber;
```
Comment (dbernatt) [inline src/main/java/com/bootcamp/demo/models/Booking.java]: serialNumber should be String type

**Review by dbernatt — CHANGES_REQUESTED** [review]

File: src/main/java/com/bootcamp/demo/models/Booking.java (hunk 1 above)
Comment (birleadrian) [reply src/main/java/com/bootcamp/demo/models/Booking.java]: We were specified not to use lombok

File: src/main/java/com/bootcamp/demo/security/SecurityConfig.java (hunk 4 above)
Comment (birleadrian) [reply src/main/java/com/bootcamp/demo/security/SecurityConfig.java]: I didn't create another user , maybe IntelliJ or something

File: src/main/java/com/bootcamp/demo/models/Booking.java (hunk 6 above)
Comment (birleadrian) [reply src/main/java/com/bootcamp/demo/models/Booking.java]: serialNumber should be a number because we will map into database

File: src/main/java/com/bootcamp/demo/models/Booking.java (hunk 2 above)
Comment (birleadrian) [reply src/main/java/com/bootcamp/demo/models/Booking.java]: I already said that we can use UUID for id , but I'm waiting for another opinion

File: src/main/java/com/bootcamp/demo/models/Booking.java (hunk 6 above)
Comment (nymphadora-tonks-team) [inline src/main/java/com/bootcamp/demo/models/Booking.java]: Highlight why you think it is necessary to have both id and serial number.

File: src/main/java/com/bootcamp/demo/models/Booking.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    private long accountId;
+    private Date startDate;
+    private Date endDate;
+    private PaymentStatus payment;
+
+    public Booking(){}
+    public Booking(long id,long serialNumber, long accountId,Date startDate,Date endDate,PaymentStatus payment){
```
Comment (nymphadora-tonks-team) [inline src/main/java/com/bootcamp/demo/models/Booking.java]: Try to ensure that the class is formatted. here for example having an empty line between the constructors and a space between `)` and `{`. That way, it will be easier improve readability.
Note that there is a keyboard shortcut which does this automatically. You can find all shortcuts here: https://resources.jetbrains.com/storage/products/intellij-idea/docs/IntelliJIDEA_ReferenceCard.pdf

File: src/main/java/com/bootcamp/demo/models/Booking.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
+package com.bootcamp.demo.models;
+import java.util.Date;
+import java.util.Objects;
+
+
+public class Booking {
+    private long id;
+    private long serialNumber;
+    private long accountId;
+    private Date startDate;
```
Comment (nymphadora-tonks-team) [inline src/main/java/com/bootcamp/demo/models/Booking.java]: Generally, it is best to avoid using this class, especially when there are other alternatives. You can find details regarding the reasons why it is best not to use it & details about a new API available starting with java 8 here: https://programminghints.com/2017/05/still-using-java-util-date-dont/, and further details about java 8 api here: https://www.baeldung.com/migrating-to-java-8-date-time-api.

File: src/main/java/com/bootcamp/demo/models/Booking.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    public Booking(long id,long serialNumber, long accountId,Date startDate,Date endDate,PaymentStatus payment){
+        this.id = id;
+        this.serialNumber = serialNumber;
+        this.accountId = accountId;
+        this.startDate = startDate;
+        this.endDate = endDate;
+        this.payment = payment;
+    }
+
+    //Getters
```
Comment (nymphadora-tonks-team) [inline src/main/java/com/bootcamp/demo/models/Booking.java]: Could you please highlight why you have added this comment?

File: src/main/java/com/bootcamp/demo/models/Booking.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        this.endDate = endDate;
+    }
+
+    public void setPayment(PaymentStatus payment) {
+        this.payment = payment;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
Comment (nymphadora-tonks-team) [inline src/main/java/com/bootcamp/demo/models/Booking.java]: I know that this method was generated as such, but it is always good to check it and see if there is anything that you would like to change.
More specifically here, what could be an issue with the fact that if is used without braces? Applies for other occurrences as well.

File: src/main/java/com/bootcamp/demo/models/Booking.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+        if (o == null || getClass() != o.getClass()) return false;
+        Booking booking = (Booking) o;
+        return id == booking.id && serialNumber == booking.serialNumber && accountId == booking.accountId && startDate.equals(booking.startDate) && endDate.equals(booking.endDate) && payment == booking.payment;
```
Comment (nymphadora-tonks-team) [inline src/main/java/com/bootcamp/demo/models/Booking.java]: - Since for the moment (at least) there is no validation for the values of the fields, what issue could occur here? Applies for other occurrences as well.
- Try to avoid lines that are very long, as they tend to be hard to follow. Try to separate these on a separate lines in a way which would be easier to follow.

File: src/main/java/com/bootcamp/demo/security/SecurityConfig.java (hunk 4 above)
Comment (nymphadora-tonks-team) [reply src/main/java/com/bootcamp/demo/security/SecurityConfig.java]: Would it be possible that instead of copy, you cut the value?

File: src/main/java/com/bootcamp/demo/models/Booking.java [hunk 12]
```diff
@@ -0,0 +1,87 @@
+package com.bootcamp.demo.models;
```
Comment (nymphadora-tonks-team) [inline src/main/java/com/bootcamp/demo/models/Booking.java]: Could you highlight why you have chosen plural for the package name?

Comment (birleadrian) [reply src/main/java/com/bootcamp/demo/models/Booking.java]: I used the plural for name , because we will have more models.

File: src/main/java/com/bootcamp/demo/models/Booking.java (hunk 6 above)
Comment (dbernatt) [reply src/main/java/com/bootcamp/demo/models/Booking.java]: But if its number than you have problems if the number is too large and don't fit in your variable, Its more secure to use String bc you don't have this kind of problems. in my pov.

File: src/main/java/com/bootcamp/demo/models/Booking.java (hunk 12 above)
Comment (dbernatt) [reply src/main/java/com/bootcamp/demo/models/Booking.java]: Here is the explanation. [link](https://softwareengineering.stackexchange.com/questions/75919/should-package-names-be-singular-or-plural)
_Use the plural for packages with homogeneous contents and the singular for packages with heterogeneous contents._

File: src/main/java/com/bootcamp/demo/models/Booking.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    public void setEndDate(LocalDateTime endDate) {
+        this.endDate = endDate;
+    }
+
+    public void setPayment(PaymentStatus payment) {
+        this.payment = payment;
+    }
+
+    @Override
+    public int hashCode() {
```
Comment (nymphadora-tonks-team) [inline src/main/java/com/bootcamp/demo/models/Booking.java]: You could have left the implementation for equals, with small updates for it to ensure that it will not cause any potential issues.

File: src/main/java/com/bootcamp/demo/models/Booking.java [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+    private UUID id;
+    private UUID serialNumber;
+    private UUID accountId;
+    private LocalDateTime startDate;
+    private LocalDateTime endDate;
+    private PaymentStatus payment;
+
+    public Booking() {}
+
+    public Booking(UUID id,UUID serialNumber, UUID accountId,LocalDateTime startDate,LocalDateTime endDate,PaymentStatus payment) {
```
Comment (nymphadora-tonks-team) [inline src/main/java/com/bootcamp/demo/models/Booking.java]: There are still some formatting inconsistencies, e.g. missing spaces after some `,`. When multiple persons collaborate on a project, it is easier to have it formatted consistently by all collaborators (for example the default formatting for intellij or a custom one can be used), as it improves readability (long term, you get accustomed to a certain formatting) & reduces the chances of being formatted only later on, when changes are made to this class.
IntelliJ keymap: https://resources.jetbrains.com/storage/products/intellij-idea/docs/IntelliJIDEA_ReferenceCard.pdf - includes also the key shortcut for "Reformat code", which will do this automatically.

File: src/main/java/com/bootcamp/demo/models/Booking.java (hunk 6 above)
Comment (nymphadora-tonks-team) [reply src/main/java/com/bootcamp/demo/models/Booking.java]: Sorry for the confusion. Forgot that for booking both an id is needed to identify it & a serialNumber to identify the scooter that was used for that booking.

File: src/main/java/com/bootcamp/demo/models/Booking.java (hunk 14 above)
Comment (dbernatt) [reply src/main/java/com/bootcamp/demo/models/Booking.java]: The serial number is a String please. change it.

**Review by dbernatt — CHANGES_REQUESTED** [review]

File: src/main/java/com/bootcamp/demo/models/Booking.java [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+    private UUID serialNumber;
+    private UUID accountId;
+    private LocalDateTime startDate;
+    private LocalDateTime endDate;
+    private PaymentStatus payment;
+
+    public Booking() {
+    }
+
+    public Booking(UUID id, UUID serialNumber, UUID accountId, LocalDateTime startDate, LocalDateTime endDate, PaymentStatus payment) {
```
Comment (dbernatt) [inline src/main/java/com/bootcamp/demo/models/Booking.java]: The booking id needs to be a random generated UUID,therefore you don't need to include to the constructor parameter list.

**Review by dbernatt — CHANGES_REQUESTED** [review]

File: src/main/java/com/bootcamp/demo/models/Booking.java (hunk 12 above)
Comment (nymphadora-tonks-team) [reply src/main/java/com/bootcamp/demo/models/Booking.java]: We could go with using singular and result in having only 1 model package.

File: src/main/java/com/bootcamp/demo/models/Booking.java (hunk 15 above)
Comment (nymphadora-tonks-team) [reply src/main/java/com/bootcamp/demo/models/Booking.java]: Also, as pointed out by @mariatirlea on slack, we could have the fields/params which are now UUID as String, and rely on other classes to do the generation as appropriate (here assign a UUID's string value). What do you think?

**Review by nymphadora-tonks-team — APPROVED** [review]
Provided that all comments are addressed before merge.
